### PR TITLE
Codable/ File System implementation - Shiva Ram Reddy

### PR DESCRIPTION
**Pull request by kathishivaramreddy**
Hey @caiozullo, @mapostolakis
Can you guys take a look on my implementation for Feed Store Challenged.Looking forward for feedback🙋‍♂️
Thanks

### CODE REVIEW

File: FeedStoreChallenge/CodableFeedStore.swift [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+				completion(error)
+			}
+		}
+	}
+	
+	public func insert(_ feed: [LocalFeedImage], timestamp: Date, completion: @escaping InsertionCompletion) {
+		let storeUrl = self.storeUrl
+		queue.async(flags: .barrier) {
+			do {
+				let encodedData = try! JSONEncoder().encode(Cache(items: feed.map { CodableFeedImage($0) } , timeStamp: timestamp))
```
Comment (caiozullo) [inline FeedStoreChallenge/CodableFeedStore.swift:71]: `try!` can lead to runtime errors. use `try` instead (if there's an error, it'll be handled in the `catch` block).

File: Tests/FeedStoreChallengeTests.swift [hunk 2]
```diff
@@ -19,84 +20,125 @@ class FeedStoreChallengeTests: XCTestCase, FeedStoreSpecs {
...
+		
+		setupMemoryBeforeTest()
+	}
+	
+	override func tearDown() {
+		super.tearDown()
+		
+		clearMemoryAfterTest()
+		
+		try? FileManager.default.removeItem(at: testSpecificUrl())
```
Comment (caiozullo) [inline Tests/FeedStoreChallengeTests.swift:34]: `clearMemoryAfterTest()` already clear the URL - so you can remove this line

File: Tests/FeedStoreChallengeTests.swift [hunk 3]
```diff
@@ -107,50 +149,59 @@ class FeedStoreChallengeTests: XCTestCase, FeedStoreSpecs {
...
-//	}
-//
-//}
+	func test_retrieve_hasNoSideEffectsOnFailure() {
+		let sut = makeSUT()
+		writeInvalidDateToMemory()
+		assertThatRetrieveHasNoSideEffectsOnFailure(on: sut)
+	}
+	
+	private func writeInvalidDateToMemory() {
```
Comment (caiozullo) [inline Tests/FeedStoreChallengeTests.swift:167]: typo: `writeInvalidDateToMemory` should be `writeInvalidDataToMemory`

File: Tests/FeedStoreChallengeTests.swift [hunk 4]
```diff
@@ -19,84 +20,125 @@ class FeedStoreChallengeTests: XCTestCase, FeedStoreSpecs {
...
+	
+	private func setupMemoryBeforeTest() {
+		removeArtifcatsFromMemory()
+	}
+	
+	private func clearMemoryAfterTest() {
+		removeArtifcatsFromMemory()
+	}
+	
+	private func removeArtifcatsFromMemory() {
```
Comment (caiozullo) [inline Tests/FeedStoreChallengeTests.swift:139]: typo: `removeArtifcatsFromMemory` should be `removeArtifactsFromMemory`

File: Tests/FeedStoreChallengeTests.swift [hunk 5]
```diff
@@ -19,84 +20,125 @@ class FeedStoreChallengeTests: XCTestCase, FeedStoreSpecs {
...
-		fatalError("Must be implemented")
+	private func makeSUT(storeUrl: URL? = nil, file: StaticString = #filePath, line: UInt = #line) -> FeedStore {
+		let sut = CodableFeedStore(storeUrl: storeUrl ?? testSpecificUrl())
+		memoryLeakTracker(instance: sut, file: file, line: line)
+		return sut
+	}
+	
+	private func memoryLeakTracker(instance: AnyObject, file: StaticString = #filePath, line: UInt = #line) {
+		addTeardownBlock { [weak instance] in
+			XCTAssertNil(instance)
```
Comment (caiozullo) [inline Tests/FeedStoreChallengeTests.swift:119]: Missing `file: file, line: line` to show messages in the right place:

```suggestion
			XCTAssertNil(instance, file: file, line: line)
```

**Review by caiozullo — COMMENTED** [review]
Hi @kathishivaramreddy, 

You almost passed the challenge!

This implementation is not an *in-memory* store because it persists data to disk.

And since it persists data, you need to implement the Integration Tests. Please, follow the instructions in the readme file.

You can also rename the PR to `Codable/File System Implementation - Shiva Ram Reddy`.

Let me know once you're done so I can review it again. Thank you 🙂

**Comment by kathishivaramreddy** [comment]
Hi @caiozullo 

I have worked on completing integration test and refactoring based on the suggestion.
After adding integration I found scope for refactoring, and have extracted duplicate code into helper method

Can you please do a re review 🙋‍♂️

File: Tests/FeedStoreHelper.swift [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+	return FileManager.default.urls(for: .cachesDirectory, in: .userDomainMask).first!
+}
+
+func testSpecificUrl(for test: AnyObject) -> URL {
+	return cachesDirectory().appendingPathComponent("\(type(of:test))-test.store")
+}
+
+func removeArtifactsFromMemory(for test: AnyObject) {
+	try? FileManager.default.removeItem(at: testSpecificUrl(for: test))
+}
```
Comment (caiozullo) [inline Tests/FeedStoreHelper.swift:29]: Those functions could be defined within an `XCTestCase` extension, so you don't need to pass the `test` parameter everywhere (you can access `self` instead)

File: Tests/FeedStoreIntegrationTests.swift [hunk 7]
```diff
@@ -17,70 +17,63 @@ class FeedStoreIntegrationTests: XCTestCase {
...
-		fatalError("Must be implemented")
-	}
-	
-	private func setupEmptyStoreState() {
-		
-	}
-	
-	private func undoStoreSideEffects() {
-		
+	private func makeSUT(storeUrl: URL? = nil, file: StaticString = #filePath, line: UInt = #line) -> FeedStore {
```
Comment (caiozullo) [inline Tests/FeedStoreIntegrationTests.swift:74]: The `storeUrl` parameter is never passed to this function. So you can remove it.

```suggestion
	private func makeSUT(file: StaticString = #filePath, line: UInt = #line) -> FeedStore {
```

**Review by caiozullo — COMMENTED** [review]
Hi @kathishivaramreddy, almost there!

The CI build failed - seems like there's a failing test.

Have a look and let me know once you're done so I can review it again. Thank you.

**Comment by kathishivaramreddy** [comment]
Hi @caiozullo 

CI is passing now, Also I have made suggested changes.

Can you please do a re review 🙋‍♂️ Thanks

**Review by caiozullo — COMMENTED** [review]
Hi @kathishivaramreddy,

Congratulations, you've passed the challenge 🥇

- All tests are passing ✅ 
- You broke down the implementation into small commits ✅ 
- Bonus points for the Integration Tests and Memory Leak Tracking ✅ 

Great job!

**Pull request closed without merging**